ISPN-13160 Write may block topology update forever

danberindei opened the pull request:
https://issues.redhat.com/browse/ISPN-13160

Remove duplicate state transfer lock in EntryWrappingInterceptor.
Acquiring a second state transfer read lock from the same
thread can now hang if another thread is waiting for
the state transfer write lock.

pruivo commented:
aren't we going to find issues if the topology changes between `EntryWrappingInterceptor` and `ClusteringDependantLogic#commitSingleEntry()`? 
I'm thinking, the `EntryWrappingInterceptor` checks the topology and it is valid but when it reaches CDL it is no longer an owner and rejects the write.

danberindei commented:
@pruivo I'm pretty sure rejecting the write without throwing an OutdatedTopologyException is not going to cause problems. Like I said, the same thing already happens in other methods, and the state transfer tests do pass (but `IracLocalStateTransferTest.testBackupRemovedKeySendAfterPrimaryFail[LON=NON_TX]` failed in CI, so maybe there's something related to IRAC).

The rebalancing process has a cluster-wide invariant: if one node's current topology has id T, then all the other nodes in the cluster must have either topology T, or T-1, or T+1 (Note: for simplicity, I'll ignore READ_NEW_WRITE_ALL, which doesn't require confirmation, but doesn't really do anything either). E.g. NO_REBALANCE doesn't require confirmation, but if a node receives a READ_OLD_WRITE_ALL(T) command before the previous NO_REBALANCE(T-1) command, it installs a fake NO_REBALANCE(T-1) topology first in order to maintain the invariant.

On the originator, `EntryWrappingInterceptor.applyChanges()` only runs after the originator received responses/acks from all the owners. 
If the topology check in `applyChanges()` still sees topology T, then it means none of the owners could have advanced to topology T+2, their commands must have finished executing in topology T or T+1.

A topology update can either add a new write owner (READ_OLD_WRITE_ALL, TRANSITORY) or remove an owner (NO_REBALANCE), not both. If topology T+1 removed a write owner, it's ok to ignore the write on that owner, because the other owners in T+1 are also owners in T and have received the command.

Although the tests pass, I think there is a problem if topology T+1 is adding owners, e.g. owners(k, T) = A and owners(k, T+1) = AB
Because I removed the check after the `commitContextEntries()` call, it is possible (although unlikely) for node A to pass the topology check at the beginning of `applyChanges()`, then install topology READ_OLD_WRITE_ALL(T+1), receive a StateTransferStartCommand from B, send the entries of k's segment to B, and only then acquire the state transfer shared lock in `ClusterDependentLogic`. That would mean k is inserted/updated on A, but it is never sent to B.

Adding back the post-commit topology check in EWI without the shared lock would work, but it would be less efficient because it would cause more retries (assuming the commit is synchronous; if it needs to commit to a blocking thread then all bets are off because the shared topology lock is released synchronously). Retrying after the entry was inserted/updated/removed will also cause linearizability violations, as another write could modify the key and cause the initial command to fail during retry.

So I'm thinking of adding back the post-commit topology check in EWI, but keep topology T for reference and only throw the OTE if topology T+1 added a new owner.

I'm also thinking of an alternative: do not let nodes request state immediately after receiving the READ_OLD_READ_NEW(T) topology, but instead make the coordinator tell the nodes they can start state transfer after all nodes have acked topology T. That actually ties with something else I've been considering for a while now, de-coupling topology updates and state transfer to allow rebalancing immediately after a node loss instead of waiting for the current state transfer to be confirmed everywhere and for old owners to be removed. It would require quite a bit of surgery in the state transfer/rebalance process so I won't do it now, I just wanted to mention it and maybe get some feedback from you.

danberindei commented:
@pruivo updated to add back the post-commit topology check, but to only retry when a new owner is added.

I also extracted a `needTopologyCheck()` from `checkTopology()`.
And I replaced [this old code](https://github.com/infinispan/infinispan/commit/d7257f7e3f4bebb95ad816aeb9cf7c192d1af6ed#diff-5ae4097194b7cc60c34de7c7698101656d0c4f4767c5c83cf72b6f5611e96d79R344) with a CACHE_MODE_LOCAL check, because I ran the core test suite and that code did nothing except for commands flagged PUT_FOR_STATE_TRANSFER|CACHE_MODE_LOCAL.

pruivo reviewed (changes requested):
forgot to press the button xD
I'll review the new changes later :+1:

danberindei commented on core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java:
```diff
@@ -659,53 +662,69 @@ public Object visitReadWriteManyEntriesCommand(InvocationContext ctx,
+                        command.hasAnyFlag(FlagBitSets.FORCE_SYNCHRONOUS);
+      if (!syncRpc)
+         return false;
+      return true;
```
I also forgot to push all the changes I was making while I was writing my comment :)

danberindei commented on core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java:
```diff
@@ -659,53 +662,69 @@ public Object visitReadWriteManyEntriesCommand(InvocationContext ctx,
-               return cs.thenRun(() -> checkTopology(ctx, command));
-            }
+      CompletionStage<Void> cs = commitContextEntries(ctx, command);
+      if (needTopologyCheck(command)) {
```
I didn't think it was worth caching the value, but `command.isSuccessful()` is not an impediment: EntryWrappingInterceptor cannot fail the command.

pruivo replied:
yeah, I thought `commitContextEntries()` could change it... my mistake

actually, why can't `applyChanges()` fail fast? 
```java
   private CompletionStage<Void> applyChanges(InvocationContext ctx, WriteCommand command) {
      if (!command.isSuccessful()) {
         return CompletableFutures.completedNull();
      }
   }
```

pruivo replied:
yeah, I thought `commitContextEntries()` could change it... my mistake

actually, why can't `applyChanges()` fail fast? 
```java
   private CompletionStage<Void> applyChanges(InvocationContext ctx, WriteCommand command) {
      if (!command.isSuccessful()) {
         return CompletableFutures.completedNull();
      }
   }
```

danberindei replied:
Sure, I can do that

danberindei replied:
Done @pruivo

pruivo reviewed (approved):
LGTM.
waiting for CI :+1:

pruivo commented:
integrated! thanks @danberindei
